src/core/linalg/src/sparse/4C_linalg_map.hpp

```diff
@@ -45,122 +64,289 @@ namespace Core::LinAlg
     /// Copy constructor from Epetra_Map
     explicit Map(const Epetra_Map& Source);
 
-    /// Copy constructor from Epetra_Map
+    /// Copy constructor from Epetra_BlockMap
     explicit Map(const Epetra_BlockMap& Source);
 
     ~Map() = default;
 
-
     Map& operator=(const Map& other);
-    void Print(std::ostream& os) const { map_->Print(os); }
 
-    //! return the reference to the Epetra_Map
-    const Epetra_Map& get_epetra_map() const { return *map_; }
-    Epetra_Map& get_epetra_map() { return *map_; }
+    //! Print object to the output stream
+    void Print(std::ostream& os) const
+    {
+      visit_variant([&](const auto& map) -> void { map.Print(os); });
+    }
+
+    //! Returns a reference of the Epetra_Map if available.
+    const Epetra_Map& get_epetra_map() const
+    {
+      auto* map = std::get_if<Utils::OwnerOrView<Epetra_Map>>(&map_);
+      if (map == nullptr)
+      {
+        FOUR_C_THROW(
+            "This Map is based on an Epetra_BlockMap, not an Epetra_Map. This cast is not "
+            "possible.");
+      }
+      return **map;
+    }
+
+    //! Returns a reference of the Epetra_Map if available.
+    Epetra_Map& get_epetra_map()
+    {
+      auto* map = std::get_if<Utils::OwnerOrView<Epetra_Map>>(&map_);
+      if (map == nullptr)
+      {
+        FOUR_C_THROW(
+            "This Map is based on an Epetra_BlockMap, not an Epetra_Map. This cast is not "
+            "possible.");
+      }
+      return **map;
+    }
+
+
+    //! Returns a const reference to the underlying Epetra_BlockMap.
+    const Epetra_BlockMap& get_epetra_block_map() const
+    {
+      // If the map_ holds directly the Epetra_BlockMap, just return it.
+      if (auto map = std::get_if<Utils::OwnerOrView<Epetra_BlockMap>>(&map_))
+      {
+        return **map;
+      }
+
+      // If the Epetra_Map is stored try to cast it.
+      auto& map = *std::get<Utils::OwnerOrView<Epetra_Map>>(map_);
+      auto* block_map = dynamic_cast<const Epetra_BlockMap*>(&map);
+
+      if (block_map == nullptr)
+      {
+        FOUR_C_THROW("The dynamic cast from Epetra_Map to Epetra_BlockMap failed.");
+      }
+
+      return *block_map;
```

Comment:
```suggestion
      return map;
```

This up-cast will always work. An `Epetra_Map` is-a `Epetra_BlockMap`. Same below.